usaspending_api/search/v2/views/spending_by_award.py

```diff
@@ -584,14 +584,70 @@ def construct_es_response_for_subawards(self, response: Response) -> dict[str, A
 
-            if row.get("Sub-Award Amount"):
-                row["Sub-Award Amount"] = float(row["Sub-Award Amount"])
+            results.append(self.calculate_complex_fields(row, hit))
 
-            row["prime_award_generated_internal_id"] = hit["unique_award_key"]
+        return self.construct_es_response(results, response)
 
-            results.append(row)
+    def calculate_complex_fields(self, row, hit):
+        if row.get("Sub-Award Amount"):
+            row["Sub-Award Amount"] = float(row["Sub-Award Amount"])
+
+        if "NAICS" in self.fields:
+            row["NAICS"] = {
+                "code": hit.get("naics"),
+                "description": hit.get("naics_description"),
+            }
+
+        if "PSC" in self.fields:
+            row["PSC"] = {
+                "code": hit.get("product_or_service_code"),
+                "description": hit.get("product_or_service_description"),
+            }
+
+        if "Assistance Listing" in self.fields:
+            row["Assistance Listing"] = {
+                "cfda_number": hit.get("cfda_number"),
+                "cfda_program_title": hit.get("cfda_titles"),
+            }
+
+        if "Sub-Recipient Location" in self.fields:
+            row["Sub-Recipient Location"] = {
+                "location_country_code": hit.get("sub_recipient_location_country_code"),
+                "country_name": hit.get("sub_recipient_location_country_name"),
+                "state_code": hit.get("sub_recipient_location_state_code"),
+                "state_name": state_name_from_code(hit.get("sub_recipient_location_state_code")),
+                "city_name": hit.get("sub_recipient_location_city_name"),
+                "county_code": hit.get("sub_recipient_location_county_code"),
+                "county_name": hit.get("sub_recipient_location_county_name"),
+                "address_line1": hit.get("sub_recipient_location_address_line1"),
+                "congressional_code": hit.get("sub_recipient_location_congressional_code"),
+                "zip4": hit.get("sub_recipient_location_zip")[-4:],
+                "zip5": hit.get("sub_recipient_location_zip5"),
+                "foreign_postal_code": hit.get("sub_recipient_location_foreign_posta"),
+            }
+
+        if "Sub-Award Primary Place of Performance" in self.fields:
+            row["Sub-Award Primary Place of Performance"] = {
+                "location_country_code": hit.get("sub_pop_country_code"),
+                "country_name": hit.get("sub_pop_country_name"),
+                "state_code": hit.get("sub_pop_state_code"),
+                "state_name": state_name_from_code(hit.get("sub_pop_state_code")),
+                "city_name": hit.get("sub_pop_city_name"),
+                "county_code": hit.get("sub_pop_county_code"),
+                "county_name": hit.get("sub_pop_county_name"),
+                "congressional_code": hit.get("sub_pop_congressional_code"),
+                "zip4": hit.get("sub_pop_zip")[-4:],
```

Comment:
Are you certain that all values in this field are either 9 characters or NULL? I only ask because I know our location data can be wonky sometimes and a change such as this could give completely wrong data if someone put a 5-digit zip into a field expecting 9 digits.

Reply:
Similar question for any of the zip related fields